sycl/include/CL/sycl/detail/cg.hpp

```diff
@@ -36,18 +36,113 @@ namespace sycl {
+// classes is an ABI break. To have an ability to have more data the following
+// approach is implemented:
+//
+// Those classes have a member - MSharedPtrStorage which is an std::vector of
+// std::shared_ptr's and is supposed to hold a reference counters of a user
+// provided shared_ptr's.
+//
+// The first element of this vector is reused to store a vector of additional
+// members handler and CG need to have.
+//
+// This additional arguments are represented using "ExtendedMemberT" structure
+// which has a pointer to an arbitrary value and an integer which is used to
+// understand how the value pointer points to should be interpreted.
+//
+// ========  ========      ========
+// |      |  |      | ...  |      | std::vector<std::shared_ptr<void>>
+// ========  ========      ========
+//    ||        ||            ||
+//    ||        \/            \/
+//    ||       user          user
+//    ||       data          data
+//    \/
+// ========  ========      ========
+// | Type |  | Type | ...  | Type | std::vector<ExtendedMemberT>
+// |      |  |      |      |      |
+// | Ptr  |  | Ptr  | ...  | Ptr  |
+// ========  ========      ========
+//
+// Prior this change this vector is supposed to have user's values only, so it
+// is not legal to expect that the first argument is a special one. Versioning
+// is implemented to overcome this problem - if first element of the
+// MSharedPtrStorage is a pointer to the special vector then CGType value has
+// version "1" encoded.
+//
+// The version of CG type is encoded in the higher byte of the value:
+//
+// 0x00000001 - CG type KERNEL version 0
+// 0x01000001 - CG type KERNEL version 1
+//   /\
+//   ||
+// The byte specifies the version
+//
+// A user of this vector should not expect that a specific data is stored at a
+// specific position, but interator over all looking for a ExtendedMemberT
+// value with desired type.
+// This allows changing/extending the content of this vector without changing
+// the version.
+//
+
+// Used to represent a type of an extended member
+enum class ExtendedMembersType: unsigned int {
+  HANDLER_KERNEL_BUNDLE = 0,
+};
+
+// Holds a pointer to an object of an arbitrary type and an ID value which
+// should be used to understand what type pointer points to.
+// Used as to extend handler class without introducing new class members which
+// would change handler layout.
+struct ExtendedMemberT {
+  ExtendedMembersType MType;
+  std::shared_ptr<void> MData;
+};
+
+static std::shared_ptr<std::vector<ExtendedMemberT>>
+convertToExtendedMembers(const std::shared_ptr<const void> &SPtr) {
+  return std::const_pointer_cast<std::vector<ExtendedMemberT>>(
+      std::static_pointer_cast<const std::vector<ExtendedMemberT>>(SPtr));
+}
+
 class stream_impl;
+class queue_impl;
+class kernel_bundle_impl;
+
+// The constant is used to shift left CG type value to access it's version
+constexpr unsigned int ShiftBitsForVersion = 24;
```

Comment:
Nit: I think this needs `static`, since it's not a function or a static member variable, and `inline` variables are not even a thing prior to C++17
```suggestion
static constexpr unsigned int ShiftBitsForVersion = 24;
```

Reply:
It seems there is no difference. constexpr variable at namepspace scope has internal linkage anyway.